xls/examples/passthrough_cocotb_test.py

```diff
@@ -0,0 +1,136 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+from copy import deepcopy
+from dataclasses import dataclass
+from pathlib import Path
+from typing import Sequence, Union
+
+import cocotb
+from cocotb.binary import BinaryValue
+from cocotb.clock import Clock
+from cocotb.handle import ModifiableObject, SimHandleBase
+from cocotb.runner import check_results_file, get_runner
+from cocotb.triggers import ClockCycles, Event, RisingEdge
+from cocotb_bus.scoreboard import Scoreboard
+from cocotb_xls import XLSChannel, XLSChannelDriver, XLSChannelMonitor
+
+from xls.common import runfiles
+
+
+@dataclass
+class SimulationData:
+  """Auxiliary structure used to store design-related data for the simulation"""
+
+  clock: Clock
+  driver: XLSChannelDriver
+  monitor: XLSChannelMonitor
+  scoreboard: Scoreboard
+  data_r: XLSChannel
+  data_s: XLSChannel
+  terminate: Event
+  clk: ModifiableObject
+  rst: ModifiableObject
+
+
+def init_sim(dut: SimHandleBase, data_to_recv: Sequence[BinaryValue]) -> SimulationData:
+  """Extracts all design-related data required for simulation"""
+
+  RECV_CHANNEL = "passthrough__data_r"
+  SEND_CHANNEL = "passthrough__data_s"
+
+  clock = Clock(dut.clk, 10, units="us")
+
+  driver = XLSChannelDriver(dut, RECV_CHANNEL, dut.clk)
+  monitor = XLSChannelMonitor(dut, SEND_CHANNEL, dut.clk)
+  data_r = XLSChannel(dut, RECV_CHANNEL)
+  data_s = XLSChannel(dut, SEND_CHANNEL)
+
+  scoreboard = Scoreboard(dut, fail_immediately=True)
+  scoreboard.add_interface(monitor, deepcopy(data_to_recv))
+
+  expected_packet_count = len(data_to_recv)
+  terminate = Event("Received the last packet of data")
+
+  def terminate_cb(_):
+    if monitor.stats.received_transactions == expected_packet_count:
+      terminate.set()
+
+  monitor.add_callback(terminate_cb)
+
+  return SimulationData(
+    clock=clock,
+    driver=driver,
+    monitor=monitor,
+    scoreboard=scoreboard,
+    data_r=data_r,
+    data_s=data_s,
+    terminate=terminate,
+    clk=dut.clk,
+    rst=dut.rst,
+  )
+
+
+@cocotb.coroutine
+async def recv(clk, send_channel):
+  """Cocotb coroutine that acts as a proc receiving data from a channel"""
+  send_channel.rdy.setimmediatevalue(0)
+  while True:
+    send_channel.rdy.value = send_channel.vld.value
+    await RisingEdge(clk)
+
+
+@cocotb.coroutine
+async def reset(clk, rst, cycles=1):
+  """Cocotb coroutine that performs the reset"""
+  rst.setimmediatevalue(1)
+  await ClockCycles(clk, cycles)
+  rst.value = 0
+
+
+@cocotb.test(timeout_time=10, timeout_unit="ms")
```

Comment:
would it make sense to rewrite this as a regular pytest test with `init_sim` as the fixture?
(similar to https://docs.cocotb.org/en/latest/runner.html)

Reply:
It is not possible because you would have to pass `dut` (a test parameter) as the argument of the `init_sim` pytest fixture. But even for simple fixtures, you really cannot mix pytest fixtures with `cocotb.test` decorators.

Invoking `test_passthrough` which is a normal, undecorated python function works when you add pytest.
Here are the changes required to launch pytest:

```diff
diff --git a/xls/examples/passthrough_cocotb_test.py b/xls/examples/passthrough_cocotb_test.py
index 4386f2f8..01025c7e 100644
--- a/xls/examples/passthrough_cocotb_test.py
+++ b/xls/examples/passthrough_cocotb_test.py
@@ -13,6 +13,7 @@
 # limitations under the License.

 import os
+import pytest
 from copy import deepcopy
 from dataclasses import dataclass
 from pathlib import Path
@@ -100,6 +101,11 @@ async def reset(clk, rst, cycles=1):
   rst.value = 0


+@pytest.fixture
+def my_fixture():
+    return "Hello from my fixture"
+
+
 @cocotb.test(timeout_time=10, timeout_unit="ms")
 async def passthrough_test(dut):
   """Cocotb test of the Passthrough proc"""
@@ -113,8 +119,8 @@ async def passthrough_test(dut):
   await sim.driver.send(test_data)
   await sim.terminate.wait()

-
-if __name__ == "__main__":
+def test_passthrough(my_fixture):
+  print(my_fixture)
   runfiles._BASE_PATH = "com_icarus_iverilog"
   iverilog_path = Path(runfiles.get_path("iverilog"))
   vvp_path = Path(runfiles.get_path("vvp"))
@@ -134,3 +140,6 @@ if __name__ == "__main__":
     waves=True,
   )
   check_results_file(results_xml)
+
+if __name__ == "__main__":
+    pytest.main(["-s", __file__])
```
After that, you will be able to observe `Hello from my fixture` printed on the log.

However if the same is applied to decorated `passthrough_test` (which is a cocotb test), the program crashes.
Here are the changes:
```diff
diff --git a/xls/examples/passthrough_cocotb_test.py b/xls/examples/passthrough_cocotb_test.py
index 4386f2f8..a1c4df33 100644
--- a/xls/examples/passthrough_cocotb_test.py
+++ b/xls/examples/passthrough_cocotb_test.py
@@ -13,6 +13,7 @@
 # limitations under the License.

 import os
+import pytest
 from copy import deepcopy
 from dataclasses import dataclass
 from pathlib import Path
@@ -100,9 +101,15 @@ async def reset(clk, rst, cycles=1):
   rst.value = 0


+@pytest.fixture
+def my_fixture():
+    return "Hello from my fixture"
+
+
 @cocotb.test(timeout_time=10, timeout_unit="ms")
-async def passthrough_test(dut):
+async def passthrough_test(dut, my_fixture):
   """Cocotb test of the Passthrough proc"""
+  print(my_fixture)
   test_data = [BinaryValue(x, n_bits=32, bigEndian=False) for x in range(10)]
   sim = init_sim(dut, test_data)

@@ -113,8 +120,7 @@ async def passthrough_test(dut):
   await sim.driver.send(test_data)
   await sim.terminate.wait()

-
-if __name__ == "__main__":
+def test_passthrough():
   runfiles._BASE_PATH = "com_icarus_iverilog"
   iverilog_path = Path(runfiles.get_path("iverilog"))
   vvp_path = Path(runfiles.get_path("vvp"))
@@ -134,3 +140,6 @@ if __name__ == "__main__":
     waves=True,
   )
   check_results_file(results_xml)
+
+if __name__ == "__main__":
+    pytest.main(["-s", __file__])
```
and the failure:
```
                                                          Cocotb test of the Passthrough proc
     0.00ns INFO     cocotb.regression                  passthrough_test failed
                                                        Traceback (most recent call last):
                                                          File "/home/rwinkler/.cache/bazel/_bazel_rwinkler/033d4de4918d187c5fea93a7c9b5eb04/sandbox/linux-sandbox/37/execroot/com_google_xls/bazel-out/k8-fastbuild/bin/xls/examples/passthrough_cocotb_test.runfiles/xls_pip_deps/pypi__cocotb/cocotb/decorators.py", line 264, in f
                                                            running_co = co(*args, **kwargs)
                                                          File "/home/rwinkler/.cache/bazel/_bazel_rwinkler/033d4de4918d187c5fea93a7c9b5eb04/sandbox/linux-sandbox/37/execroot/com_google_xls/bazel-out/k8-fastbuild/bin/xls/examples/passthrough_cocotb_test.runfiles/xls_pip_deps/pypi__cocotb/cocotb/decorators.py", line 77, in __call__
                                                            return _RunningCoroutine(self._func(*args, **kwargs), self)
                                                        TypeError: passthrough_test() missing 1 required positional argument: 'my_fixture'
     0.00ns INFO     cocotb.regression                  **************************************************************************************************
                                                        ** TEST                                      STATUS  SIM TIME (ns)  REAL TIME (s)  RATIO (ns/s) **
                                                        **************************************************************************************************
                                                        ** passthrough_cocotb_test.passthrough_test   FAIL           0.00           0.00          2.79  **
                                                        **************************************************************************************************
                                                        ** TESTS=1 PASS=0 FAIL=1 SKIP=0                              0.00           0.12          0.01  **
                                                        **************************************************************************************************
```

Reply:
Because `pytest` can be used only for parametrizing a call to Cocotb Runner, I wonder whether it can be useful for us.

Reply:
> However if the same is applied to decorated passthrough_test (which is a cocotb test), the program crashes.

Was this limitation already reported upstream?

Reply:
that particular comment seem to hint at a testing style that could support having the `dut` being passed as a fixture.

Reply:
This limitation is known, there are open issues related to tight integration with pytests and this feature is still a work in progress. In fact, it was never stated that cocotb tests can be parameterized with pytest. Here are the relevant issues and PRs:
* https://github.com/cocotb/cocotb/issues/494
* https://github.com/cocotb/cocotb/issues/791
* https://github.com/cocotb/cocotb/pull/1782

I asked for this feature one more time in the issue: https://github.com/cocotb/cocotb/issues/494#issuecomment-1750147747